src/lightning/data/cache/dataset_optimizer.py

```diff
@@ -0,0 +1,761 @@
+import logging
+import os
+import signal
+import traceback
+from abc import ABC, abstractmethod
+from enum import Enum
+from multiprocessing import Process, Queue
+from pathlib import Path
+from queue import Empty
+from shutil import copyfile
+from threading import Thread
+from time import sleep, time
+from typing import Any, Callable, Dict, List, Literal, Optional, Tuple
+from urllib import parse
+
+from tqdm import tqdm
+
+from lightning import seed_everything
+from lightning.data.cache import Cache
+from lightning.data.cache.constants import (
+    _BOTO3_AVAILABLE,
+    _DEFAULT_FAST_DEV_RUN_ITEMS,
+    _INDEX_FILENAME,
+    _LIGHTNING_CLOUD_GREATER_EQUAL_0_5_41,
+    _TORCH_GREATER_EQUAL_2_1_0,
+)
+
+if _TORCH_GREATER_EQUAL_2_1_0:
+    from torch.utils._pytree import tree_flatten, tree_unflatten
+
+if _LIGHTNING_CLOUD_GREATER_EQUAL_0_5_41:
+    from lightning_cloud.resolver import _LightningSrcResolver, _LightningTargetResolver
+
+if _BOTO3_AVAILABLE:
+    import boto3
+
+logger = logging.Logger(__name__)
+
+
+def _get_cache_folder() -> str:
+    """Returns the cache folder."""
+    return os.getenv("CACHE_FOLDER", "/cache")
+
+
+def _get_num_nodes() -> int:
+    """Returns the number of nodes."""
+    return int(os.getenv("NUM_NODES", 1))
+
+
+def _get_node_rank() -> int:
+    """Returns the current node rank of the instance."""
+    return int(os.getenv("NODE_RANK", 0))
+
+
+def _get_fast_dev_run() -> int:
+    """Returns whether fast dev mode is enabled."""
+    return bool(int(os.getenv("FAST_DEV_RUN", 1)))
+
+
+def _get_home_folder() -> str:
+    """Returns whether cache folder for the filepaths."""
+    return os.getenv("HOME_FOLDER", os.path.expanduser("~"))
+
+
+def _download_data_target(src_dir: str, remote_src_dir: str, cache_dir: str, queue_in: Queue, queue_out: Queue) -> None:
```

Comment:
These three functions have a lot of different code paths. I think a test for each is needed here.

Reply:
The test won't pass if the block weren't tested but sure.